From 8d1687f42bf86fe8197950b86a50936e28054534 Mon Sep 17 00:00:00 2001
From: Andre Przywara <andre.przywara@arm.com>
Date: Sat, 10 Dec 2016 12:21:38 +0000
Subject: [PATCH] sunxi: power: allow setting CPU voltage

Provide high level functions to allow programming certain voltages for
the CPU cores. This calculates the respective AXP803 register value for
the DCDC2/DCDC3 pair from a given voltage.
Also this allow to turn the CPU cores off completely, to easily
implement a shutdown functionality.

Signed-off-by: Andre Przywara <andre.przywara@arm.com>
---
 plat/sun50iw1p1/sunxi_power.c   | 30 ++++++++++++++++++++++++++++++
 plat/sun50iw1p1/sunxi_private.h |  2 ++
 2 files changed, 32 insertions(+)

diff --git a/plat/sun50iw1p1/sunxi_power.c b/plat/sun50iw1p1/sunxi_power.c
index 0c2487e20..095a0e757 100644
--- a/plat/sun50iw1p1/sunxi_power.c
+++ b/plat/sun50iw1p1/sunxi_power.c
@@ -176,6 +176,31 @@ static void rsb_wait(const char *desc)
 	ERROR("%s: 0x%x\n", desc, reg);
 }
 
+static int axp803_set_cpu_voltage(int millivolt)
+{
+	uint8_t reg;
+
+	if (millivolt <= 0) {			/* power off system */
+		sunxi_pmic_write(0x32, sunxi_pmic_read(0x32) | 0x80);
+		return 0;			/* hopefully not ... */
+	}
+
+	if (millivolt < 800 || millivolt > 1300)
+		return -1;
+
+	if (millivolt > 1200)
+		reg = (millivolt - 1200) / 20 + 70;
+	else
+		reg = (millivolt - 500) / 10 + 0;
+
+	sunxi_pmic_write(0x21, reg);	/* DCDC2 */
+
+	while (!(sunxi_pmic_read(0x21) & 0x80))
+		;
+
+	return 0;
+}
+
 /* Initialize the RSB PMIC connection. */
 static int pmic_init(uint16_t hw_addr, uint8_t rt_addr)
 {
@@ -266,6 +291,11 @@ static int pmic_setup(void)
  
 	sunxi_pmic_write(0x15, 0x1a);	/* DLDO1 = VCC3V3_HDMI voltage = 3.3V */
 
+	ret = sunxi_rsb_read(0x14);
+	sunxi_rsb_write(0x14, ret | 0x40);	/* DCDC2/3 dual phase */
+
+	axp803_set_cpu_voltage(1100);
+
 	return 0;
 }
 
diff --git a/plat/sun50iw1p1/sunxi_private.h b/plat/sun50iw1p1/sunxi_private.h
index 519422343..0c8b640bd 100644
--- a/plat/sun50iw1p1/sunxi_private.h
+++ b/plat/sun50iw1p1/sunxi_private.h
@@ -72,6 +72,8 @@ int sunxi_pmic_setup(void);
 int sunxi_pmic_read(uint8_t address);
 int sunxi_pmic_write(uint8_t address, uint8_t value);
 
+int sunxi_power_set_cpu_voltage(int millivolt);
+
 void udelay(unsigned int delay);
 int sunxi_setup_clocks(uint16_t socid);
 
